from keras.callbacks import ModelCheckpoint


class MyModelCheckpoint(ModelCheckpoint):
    """This is a workaround to fix the problem that ModelCheckpoint cannot save the model created by keras.utils.multi_gpu_model.
    # Arguments
        model: A Keras model instance except the model generated by multi_gpu_model.
    # References
        https://www.twblogs.net/a/5b8119bd2b71772165ab0275
    """

    def __init__(self, model, filepath, monitor='val_loss', verbose=0,
                 save_best_only=False, save_weights_only=False,
                 mode='auto', period=1):
        self.model_to_save = model
        super().__init__(filepath, monitor, verbose,
                         save_best_only, save_weights_only, mode, period)

    def set_model(self, model):
        super().set_model(self.model_to_save)
